pip/pip-297.md

```diff
@@ -0,0 +1,168 @@
+
+**Function Framework** is a framework for running the Function instance.
+
+**Function Context** is an interface that provides access to various information and resources for the connector or the
+function. The function context is passed to the connector or the function when it is initialized, and then can be used
+to interact with the Pulsar system.
+
+The current implementation of the exception handler:
+**Function instance thread**: The function framework initializes a thread for each function instance to handle the
+core logic of the function/connector, including consuming messages from the Pulsar topic for the sink connector,
+executing the logic of the function, producing messages to the Pulsar topic for the source connector, handling the
+exception, etc.
+
+**Exception handling logic**: The function itself can throw exceptions, and this thread will catch the exception and
+then close the function. This means that the function will stop working until it is restarted manually or
+automatically by the function framework.
+
+Even though it is not explicitly defined, there are two types of exceptions that could be handled for the function or
+the framework.
+
+- **Recoverable exception**: This is an exception that the function can recover from by itself, such as network
+  connection issues, persistence issues, etc. The function can catch these exceptions and retry the operation until it
+  succeeds or reaches a limit.
+- **Unrecoverable exception**: This is an exception that the function cannot recover from by itself and needs to notify
+  the framework to terminate it. These are fatal exceptions that indicate a configuration issue, a logic error, or an
+  incompatible system. The function framework will catch these exceptions, report them to users, and terminate the
+  function.
+
+# Motivation
+
+The current implementation of the connector and Pulsar Function exception handler cannot handle the unrecoverable
+exceptions that are thrown outside of the function instance thread.
+
+For example, suppose we have a sink connector that uses its own threads to batch-sink the data to an external system. If
+any unrecoverable exceptions occur in those threads, the function instance thread will not be aware of them and will
+not be able to terminate the connector. This will cause the connector to hang indefinitely. There is a related issue
+here: https://github.com/apache/pulsar/issues/9464
+
+The same problem exists for the source connector. The source connector may also use a separate thread to fetch data from
+an external system. If any fatal exceptions happen in that thread, the connector will also hang forever. This issue has
+been observed for the Kafka source connector: https://github.com/apache/pulsar/issues/9464. We have fixed it by adding
+the notifyError method to the PushSource class in PIP-281: https://github.com/apache/pulsar/pull/20807. However, this
+does not solve the same problem that all source connectors face.
```

Comment:
Why the notifyError method can't solve the same problem for all source connectors? It's not a specific fix for Kafka connector, right?

Reply:
I added the explanation to the PIP.

The PIP-281 only affects `PushSource`. Not all source connectors are implemented based on `PushSource`, such as debezium connector.